lib/common/pool.h

```diff
@@ -0,0 +1,45 @@
+/**
+ * Copyright (c) 2016-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+#ifndef POOL_H
+#define POOL_H
+
+#include <stddef.h>   /* size_t */
+
+typedef struct POOL_ctx_s POOL_ctx;
+
+/*! POOL_create() :
+    Create a thread pool with at most `numThreads` threads.
+    `numThreads` must be at least 1.
+    The maximum number of queued jobs before blocking is `queueSize`.
+    `queueSize` must be at least 1.
+    @return : The POOL_ctx pointer on success else NULL.
+*/
+POOL_ctx *POOL_create(size_t numThreads, size_t queueSize);
+
+/*! POOL_free() :
+    Free a thread pool returned by POOL_create().
+*/
+void POOL_free(POOL_ctx *ctx);
+
+/*! POOL_function :
+    The function type that can be added to a thread pool.
+*/
+typedef void (*POOL_function)(void *);
+/*! POOL_add_function :
+    The function type for a generic thread pool add function.
+*/
+typedef void (*POOL_add_function)(void *, POOL_function, void *);
+
+/*! POOL_add() :
+    Add the job `function(opaque)` to the thread pool.
+    Possibly blocks until there is room in the queue.
+*/
+void POOL_add(void *ctx, POOL_function function, void *opaque);
```

Comment:
I selected the `void*` version because it can be used as an argument to a function with a signature like:

    /* Function takes a thread pool to run jobs on */
    void fn(void *threadPool, POOL_add_function addJob) {
      addJob(threadPool, &someFn, opaque);
    }

I originally had `POOL_add()` and `POOL_add_generic()`.  I combined them because the linker on the `make aarch64test` [segfaulted](https://travis-ci.org/terrelln/zstd/jobs/187623333) on `POOL_add_generic()`.  I picked the more generic one so I could get the tests passing.  If they can both be added without segfaulting the linker, that would be great.